### ISSUE TITLE
add `allowedPattern` in `quote-props`

### ISSUE THREAD

**Issue by XadillaX** [eslint/eslint#8220]
**What rule do you want to change?**

`quote-props`

**Does this change cause the rule to produce more or fewer warnings?**

I want anything not in camel case to be quote-wrapped in object definition.

```js
{
    appId: ...,
    "content-type": ...,
    "underscore_thing": ...
}
```

Because I regard any key in object as a variable name and variable name should be camel case. If it's not in camel case, it should be wrapped with quotes so that I can regard it as a string.

In fact, I have another configuration in read:

```javascript
// enforce dot notation whenever possible
"dot-notation": [ "error", {
    allowPattern: "([a-zA-Z0-9])(_)+([a-zA-Z0-9])"
}],
```

That mean I can write code like:

```javascript
temp.aAndB = 1;
temp._a_ = 2;
temp["a_1"] = 3;
temp["a-1"] = abcdeft();
temp["__a_____b__"] = .4;
```

**How will the change be implemented? (New option, new default behavior, etc.)?**

Two ways:

+ Regard underscored key as `quote-needed` so that it won't occur warnings.
+ Add an options `allowPattern` just like `dot-notation` do.

**Please provide some example code that this change will affect:**

```js
{
    appId: ...,
    "underscore_thing": ...
}
```

**What does the rule currently do for this code?**

Unnecessarily quoted property found.

**What will the rule do after it's changed?**

No warning.

---------------------------

> ## The previous version of this issue
>
> **Tell us about your environment**
> 
> * **ESLint Version:** 3.16.1
> * **Node Version:** 6.9.1
> * **npm Version:** 2.16.11
> 
> **What parser (default, Babel-ESLint, etc.) are you using?**
> 
> default.
> 
> **Please show your full configuration:**
> 
> https://github.com/XadillaX/eslint-config-xadillax-style
> 
> **What did you do? Please include the actual source code causing the issue.**
> 
> ```js
>         const query = {
>             appid: akyuu.config.appId,
>             "redirect_uri": `${akyuu.config.server.baseUrl}/signin`,
>             "response_type": "code"
>         };
> ```
> 
> **What did you expect to happen?**
> 
> I want anything not in camel case to be quote-wrapped in object definition.
> 
> ```js
> {
>     appId: ...,
>     "content-type": ...,
>     "underscore_thing": ...
> }
> ```
> 
> Because I regard any key in object as a variable name and variable name should be camel case. If it's not in camel case, it should be wrapped with quotes so that I can regard it as a string.
> 
> In fact, I have another configuration in read:
> 
> ```javascript
>         // enforce dot notation whenever possible
>         "dot-notation": [ "error", {
>             allowPattern: "([a-zA-Z0-9])(_)+([a-zA-Z0-9])"
>         }],
> ```
> 
> That mean I can write code like:
> 
> ```javascript
> temp.aAndB = 1;
> temp._a_ = 2;
> temp["a_1"] = 3;
> temp["a-1"] = abcdeft();
> temp["__a_____b__"] = .4;
> ```
> 
> **What actually happened? Please include the actual, raw output from ESLint.**
> 
> When I wrapped with quotes, it shows me `Unnecessarily quoted property found`. And when I removed the quotes, it shows me `Identifier is not in camel case`.
> 
> So I want to wrap it with quotes and let ESlint judge it's needed.

**Comment by kaicataldo** [comment]
Would you mind filling out the [rule change proposal](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md) template instead of the bug reporting one? It sounds like you'd like to enhance this rule with a new option but I'm not sure entirely sure what you're envisioning for the API.

**Comment by XadillaX** [comment]
@kaicataldo I'v updated this issue.

**Comment by XadillaX** [comment]
anyone who to deal with this issue?

**Comment by XadillaX** [comment]
I CAN'T LINT MY CODE NOW. ANYONE WHO TO DEAL WITH THIS ISSUE?

**Comment by platinumazure** [comment]
Hi @XadillaX, thanks for writing an issue. First off, on behalf of the team, I want to apologize for letting this sit so long- that was unintentional and not something we want to happen.

That said, @kaicataldo [had requested](https://github.com/eslint/eslint/issues/8220#issuecomment-285481192) that the issue be updated to conform to our rule change proposal, and I haven't seen that happen yet. So unfortunately there isn't much we can do at this point with this issue, because we're not clear on what rule(s) should change and in what way to satisfy your use case. (Disclaimer: This does not constitute a promise that the ESLint team will accept your rule change proposal; in that case, you can create a custom rule that will do what you want.)

**Comment by ilyavolodin** [comment]
@platinumazure Issue has been updated. Sort-of.
> Because I regard any key in object as a variable name and variable name should be camel case. If it's not in camel case, it should be wrapped with quotes so that I can regard it as a string.
This seems like an extremely personal preference. I can't say that I've ever seen anyone doing that. So I don't think this would be a good fit for the core. I would suggest that you copy the rule and create your own implementation as a plugin.

**Comment by XadillaX** [comment]
@ilyavolodin the problem is:

When I remove quotes, it appears:

```
Identifier '******' is not in camel case.
```

And when I add quotes, it appears:

```
Unnecessarily quoted property '******' found.
```

So I want some underlined keys in `Object` be valid.

**Comment by not-an-aardvark** [comment]
I think the solution for that issue would be to not enforce that property names must be camelcase. You can do that by setting the `properties: never` option in the [`camelcase`](http://eslint.org/docs/rules/camelcase) rule.

**Comment by not-an-aardvark** [comment]
Based on the number of :+1:s on https://github.com/eslint/eslint/issues/8220#issuecomment-289287241, it seems like there's consensus that this option would not be a good fit for core, so I'm closing this issue.
